tools/generate_tables_sql_and_testNames.php

```diff
@@ -1,118 +1,234 @@
+                } else if ($bits[0]=="date") {
+                    if (empty($bits[3])
+                        || empty($bits[4])
+                    ) {
+                        $error_message = "Instrument(s) file has invalid "
+                            . "instrument data.";
+                    }
+                } else {
+                    // Safety SQL injection prevention.
+                    $bits[0] = str_replace('`', '', $db->escape($bits[0]));
                 }
-
-                $bits[2]=htmlspecialchars($bits[2]);
-                $output.="`$bits[1]` $bits[0] default NULL,\n";
+                $bits[2]     = htmlspecialchars($bits[2]);
+                $output     .= "`$bits[1]` $bits[0] default NULL,\n";
+                $column_name = str_replace('`', '', $db->escape($bits[1]));
+                $sql_query_columns[] = "`$column_name` $bits[0] default NULL,\n";
+            }
         }
-
+        $output .= "PRIMARY KEY  (`CommentID`)\n) ENGINE=InnoDB"
+                  ." DEFAULT CHARSET=utf8;\n";
+        $fp      = fopen($file_name, "w");
+        fwrite($fp, $output);
+        // Check if Table Name exists
+        if (empty($sql_query_table_name)) {
+            $error_message = "Instrument(s) file has invalid instrument data.";
+        } else {
+            fwrite(
+                $fp,
+                "REPLACE INTO test_names (Test_name, Full_name, Sub_group) "
+                ."VALUES ('$sql_query_table_name', '$title', 1);\n"
+            );
+            // Writes a record of the changes to a file but is not sourced directly.
+            foreach ($pages as $page_number => $page) {
+                fwrite(
+                    $fp,
+                    "INSERT INTO instrument_subtests "
+                    ."(Test_name, Subtest_name, Description)"
+                    ." VALUES ('$sql_query_table_name', '"
+                    .$sql_query_table_name."_page".($page_number+1)
+                    ."', '$page');\n"
+                );
+            }
+        }
+        // Writes the error message to the sql file.
+        if (!empty($error_message)) {
+            fwrite(
+                $fp,
+                "/* The SQL file is invalid! \n"
+                ."tools/generate_tables_sql_and_testNames.php error_message:\n"
+                ."$error_message\n"
+                ."*/"
+            );
+        }
+        fclose($fp);
     }
-    $output.="PRIMARY KEY  (`CommentID`)\n) ENGINE=InnoDB DEFAULT CHARSET=utf8;\n";
-    print "Filename: $filename\n";
-    $fp=fopen($filename, "w");
-    fwrite($fp, $output);
-    fwrite($fp, "REPLACE INTO test_names (Test_name, Full_name, Sub_group) VALUES ('$tablename', '$title', 1);\n");
-    foreach($pages as $pageNo => $page) {
-        fwrite($fp, "INSERT INTO instrument_subtests (Test_name, Subtest_name, Description) VALUES ('$tablename', '" . $tablename  . "_page" . ($pageNo+1) . "', '$page');\n");
 
+    // Execute the SQL tasks if no error_message.
```

Comment:
Why is anything being executed here at all?

Reply:
@ZainVirani not sure if I understand what you mean totally, but this is a rare example of a tool that is called by the front-end and commands are expected to be run. 
We could add an argument here but I imagine it would always be passed when called from the front-end

Reply:
this script can also be run from command line, and the dev may not want the queries to be run

Reply:
Hi @johnsaigle and @ZainVirani the latest commit adds the feature of not executing the SQL query if a secondary argument "false" is supplied. See the comment examples in the generate_tables_sql_and_testNames.php file for how it works.